File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/logic/BusTripServiceFilter.java
package com.example.prj_gifu_univ_bus_navi.logic;

import com.example.prj_gifu_univ_bus_navi.model.BusTrip;
import com.example.prj_gifu_univ_bus_navi.model.OperationRule;
import com.example.prj_gifu_univ_bus_navi.model.ServiceDateContext;
import java.util.ArrayList;
import java.util.List;

public final class BusTripServiceFilter {
    private BusTripServiceFilter() {
    }

    public static boolean isTripAvailable(BusTrip trip, ServiceDateContext context) {
        if (trip.getBaseDayType() != context.getBaseDayType()) {
            return false;
        }
        OperationRule rule = trip.getOperationRule();
        switch (rule) {
            case NONE:
                return true;
            case SCHOOL_HOLIDAY_EXCLUDED:
                return !context.isSchoolHoliday();
            case SCHOOL_HOLIDAY_ONLY:
                return context.isSchoolHoliday();
            case LIMITED_PERIOD:
                return isInOperatingMonth(trip, context.getMonth());
            case LIMITED_PERIOD_AND_SCHOOL_HOLIDAY_EXCLUDED:
                return isInOperatingMonth(trip, context.getMonth()) && !context.isSchoolHoliday();
            default:
                return false;
        }
    }

    public static List<BusTrip> filterAvailableTrips(List<BusTrip> trips, ServiceDateContext context) {
        List<BusTrip> available = new ArrayList<>();
        for (BusTrip trip : trips) {
            if (isTripAvailable(trip, context)) {
                available.add(trip);
            }
        }
        return available;
    }

    private static boolean isInOperatingMonth(BusTrip trip, int month) {
        Integer start = trip.getOperatingStartMonth();
        Integer end = trip.getOperatingEndMonth();
        if (start == null || end == null) {
            return false;
        }
        if (start <= end) {
            return month >= start && month <= end;
        }
        return month >= start || month <= end;
    }
}

File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/logic/BusTripServiceFilter.kt
package com.example.prj_gifu_univ_bus_navi.logic

import com.example.prj_gifu_univ_bus_navi.model.BusTrip
import com.example.prj_gifu_univ_bus_navi.model.OperationRule
import com.example.prj_gifu_univ_bus_navi.model.ServiceDateContext

object BusTripServiceFilter {
    fun isTripAvailable(trip: BusTrip, context: ServiceDateContext): Boolean {
        if (trip.baseDayType != context.baseDayType) return false
        return when (trip.operationRule) {
            OperationRule.NONE -> true
            OperationRule.SCHOOL_HOLIDAY_EXCLUDED -> !context.isSchoolHoliday
            OperationRule.SCHOOL_HOLIDAY_ONLY -> context.isSchoolHoliday
            OperationRule.LIMITED_PERIOD -> isInOperatingMonth(trip, context.month)
            OperationRule.LIMITED_PERIOD_AND_SCHOOL_HOLIDAY_EXCLUDED -> {
                isInOperatingMonth(trip, context.month) && !context.isSchoolHoliday
            }
        }
    }

    fun filterAvailableTrips(trips: List<BusTrip>, context: ServiceDateContext): List<BusTrip> =
        trips.filter { isTripAvailable(it, context) }

    private fun isInOperatingMonth(trip: BusTrip, month: Int): Boolean {
        val start = trip.operatingStartMonth ?: return false
        val end = trip.operatingEndMonth ?: return false
        return if (start <= end) {
            month in start..end
        } else {
            month >= start || month <= end
        }
    }
}
